File: DigitalImageProcessing/cpp/local_threshold_hist.cpp
/**
 * @File    : local_threshold_hist.cpp
 * @Brief   : 局部阈值分割；直方图技术
 * @Author  : Wei Li
 * @Date    : 2021-09-28
*/

#include <iostream>
#include <cmath>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

cv::Mat calcGrayHist(const cv::Mat &image)
{
    cv::Mat histogram = cv::Mat::zeros(cv::Size(256, 1), CV_32SC1);
    int rows = image.rows;
    int cols = image.cols;
    // 统计灰度等级数量
    for (size_t idx_row = 0; idx_row < rows; ++idx_row)
    {
        for (size_t idx_col = 0; idx_col < cols; ++idx_col)
        {
            int index = int(image.at<uchar>(idx_row, idx_col));
            histogram.at<int>(0, index) += 1;
        }
    }
    return histogram;
}

int threshTwoPeaks(const cv::Mat &image, cv::Mat &thresholdImg)
{
    cv::Mat histogram = calcGrayHist(image);
    // 找到灰度直方图最大峰值对应的灰度值
    cv::Point firstPeakLoc;
    cv::minMaxLoc(histogram, nullptr, nullptr, nullptr, &firstPeakLoc);
    int firstPeak = firstPeakLoc.x;
    // 寻找灰度直方图第二个峰值对应的灰度值
    cv::Mat measureDists = cv::Mat::zeros(cv::Size(256, 1), CV_32SC1);
    for (int k = 0; k < 256; ++k)
    {
        int hist_k = histogram.at<int>(0, k);
        measureDists.at<float>(0, k) = std::pow(float(k - firstPeak), 2)*hist_k;
    }
    cv::Point secondPeakLoc;
    cv::minMaxLoc(measureDists, nullptr, nullptr, nullptr, &secondPeakLoc);
    int secondPeak = secondPeakLoc.x;
    // 找到两个峰值之间最小值对应的灰度值，作为阈值
    cv::Point thresholdLoc;
    int thresholdVal = 0;
    // 第一个峰值在第二个峰值的左侧
    if (firstPeak < secondPeak)
    {
        cv::minMaxLoc(histogram.colRange(firstPeak, secondPeak), nullptr, nullptr, &thresholdLoc);
        thresholdVal = firstPeak + thresholdLoc.x + 1;
    }
    else // 第一个峰值在第二个峰值的右侧
    {
        cv::minMaxLoc(histogram.colRange(secondPeak, firstPeak), nullptr, nullptr, &thresholdLoc);
        thresholdVal =secondPeak + thresholdLoc.x + 1;
    }

    // 进行阈值分割
    cv::threshold(image, thresholdImg, thresholdVal, 255, cv::THRESH_BINARY);
    return thresholdVal;
}

// -------------------------------
int main(int argc, char **argv)
{
    if (argc > 1)
    {
        cv::Mat image = cv::imread(argv[1], 0);
        if (!image.data)
        {
            std::cout << "Error: reading image unsuccesfully." << std::endl;
            return -1;
        }
        cv::imshow("OriginImage", image);

        cv::Mat thresholdImg;
        int thresholdVal = threshTwoPeaks(image, thresholdImg);
        std::cout << "The threshold value is " << thresholdVal << std::endl;
        cv::imshow("ThresholdImage", thresholdImg);

        cv::waitKey(0);
        cv::destroyAllWindows();
        return 0;
    }
    else
    {
        std::cout << "Usage: OpenCV gauss convolution image." << std::endl;
        return -1;
    }
}

File: DigitalImageProcessing/python/local_threshold_hist.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 局部阈值分割; 直方图技术；
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-09-28
"""

import sys
import numpy as np
import cv2 as cv


def calcGrayHist(image):
    rows, cols = image.shape
    grayHist = np.zeros([256], np.uint64)
    for idx_row in range(rows):
        for idx_col in range(cols):
            grayHist[image[idx_row][idx_col]] += 1
    
    return grayHist

def threshTwoPeaks(image):
    histogram = calcGrayHist(image)
    # 找到灰度直方图的最大峰值对应的灰度值
    maxLoc = np.where(histogram==np.max(histogram))
    firstPeak = maxLoc[0][0]
    # 寻找灰度直方图的第二个峰值对应的灰度图
    measureDists = np.zeros([256], np.float32)
    for k in range(256):
        measureDists[k] = pow(k - firstPeak,  2) * histogram[k]
    maxLoc2 = np.where(measureDists==np.max(measureDists))
    secondPeak = maxLoc2[0][0]
    # 找到两个峰值之间的最小值对应的灰度值,作为阈值
    threshold  = 0
    # 第一个峰值在第二个峰值的右侧
    if firstPeak > secondPeak:
        temp = histogram[int(secondPeak):int(firstPeak)]
        minLoc = np.where(temp==np.min(temp))
        threshold = secondPeak + minLoc[0][0] + 1
    # 第一个峰值在第二个峰值的左侧
    else:
        temp = histogram[int(firstPeak):int(secondPeak)]
        minLoc = np.where(temp==np.min(temp))
        threshold = firstPeak + minLoc[0][0] + 1

    # 找到阈值后进行阈值处理，获得二值图
    thresholdImg = image.copy()
    thresholdImg[thresholdImg > threshold] = 255
    thresholdImg[thresholdImg <= threshold] = 0
    return (threshold, thresholdImg)


# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or directory.")
            sys.exit()
        cv.imshow("OriginImage", image)

        thresholdVal, thresholdImg = threshTwoPeaks(image)
        print(f"The threshold value is {thresholdVal}")
        cv.imshow("ThresholdImage", thresholdImg)

        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print(f"Usage: python gaussBlur imageFile.")
